File: 0108-Convert Sorted Array to Binary Search Tree/c++/solution.cpp
#include <bits/stdc++.h>
using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        if(nums.empty()) return nullptr;
        return helper(nums, 0, nums.size()-1);
    }
    
    TreeNode* helper(vector<int>& nums, int start_index, int end_index) {
        if(start_index > end_index) return nullptr;
        int middle_index = (start_index + end_index)/2;
        TreeNode *middle = new TreeNode(nums[middle_index]);
        middle->left = helper(nums, start_index, middle_index-1);
        middle->right = helper(nums, middle_index+1, end_index);
        return middle;
    }
};

class Solution {
public:
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        if(nums.empty()) return nullptr;
        int middle_index = nums.size()/2;
        auto begin_iter = nums.begin();
        TreeNode *middle = new TreeNode(nums[middle_index]);
        vector<int> left(begin_iter, begin_iter+middle_index);
        vector<int> right(begin_iter+middle_index+1, nums.end());
        middle->left = sortedArrayToBST(left);
        middle->right = sortedArrayToBST(right);
        return middle;
    }
};

// Runtime: 28 ms, faster than 34.62% of C++ online submissions for Convert Sorted Array to Binary Search Tree.
// Memory Usage: 25.9 MB, less than 15.19% of C++ online submissions for Convert Sorted Array to Binary Search Tree.

File: 0108-Convert Sorted Array to Binary Search Tree/golang/solution.go
package golang

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func sortedArrayToBST(nums []int) *TreeNode {
	length := len(nums)
	if length == 0 {
		return nil
	}
	middle := length / 2
	return &TreeNode{
		Val:   nums[middle],
		Left:  sortedArrayToBST(nums[:middle]),
		Right: sortedArrayToBST(nums[middle+1:]),
	}
}

// Runtime: 88 ms, faster than 98.32% of Go online submissions for Convert Sorted Array to Binary Search Tree.
// Memory Usage: 247.7 MB, less than 5.33% of Go online submissions for Convert Sorted Array to Binary Search Tree.
